### PR TITLE
Slotted Cartesian Product

### PR DESCRIPTION
**Pull request by alexaverbuch**
Branched from https://github.com/neo4j/neo4j/pull/9883

### CODE REVIEW

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala [hunk 1]
```diff
@@ -29,6 +29,26 @@ import scala.collection.mutable.ArrayBuffer
 
 class MatchAcceptanceTest extends ExecutionEngineFunSuite with QueryStatisticsTestSupport with CypherComparisonSupport {
 
+
+  test("apa") {
```
Comment (craigtaverner) [inline enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala]: Is this a test to throw away?

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala [hunk 2]
```diff
@@ -487,10 +507,10 @@ return p""")
         |RETURN host""".stripMargin
 
     //WHEN
-    val result = profile(query)
-
+    val result = succeedWith(Configs.Interpreted, query)
     //THEN
-    result.toList should equal(List(Map("host" -> host), Map("host" -> null)))
+    println(result.getExecutionPlanDescription)
```
Comment (craigtaverner) [inline enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala]: Temp code?

File: enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/PrimitiveExecutionContext.scala [hunk 3]
```diff
@@ -24,29 +24,33 @@ import org.neo4j.cypher.internal.frontend.v3_3.InternalException
...
+        throw new InternalException("Tried to copy more data into less.")
+      else {
+        System.arraycopy(longs, 0, other.longs, longOffset, pipeline.numberOfLongs)
+        System.arraycopy(refs, 0, other.refs, refOffset, pipeline.numberOfReferences)
+      }
     case _ => fail()
   }
 
-  def copyFrom(input: ExecutionContext): Unit = input match {
+  def copyFrom(input: ExecutionContext, nLongs: Int, nRefs: Int): Unit = input match {
```
Comment (craigtaverner) [inline enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/PrimitiveExecutionContext.scala]: Missing override?

**Review by craigtaverner — CHANGES_REQUESTED** [review]

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala (hunk 1 above)
Comment (alexaverbuch) [reply enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala]: It's not from this PR

File: enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/PrimitiveExecutionContext.scala (hunk 3 above)
Comment (alexaverbuch) [reply enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/PrimitiveExecutionContext.scala]: @craigtaverner I think all review comments are now addressed

File: enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/EnterprisePipeBuilder.scala [hunk 4]
```diff
@@ -231,12 +231,23 @@ class EnterprisePipeBuilder(fallback: PipeBuilder,
...
 
+      case _: SemiApply |
+           _: AntiSemiApply =>
+        fallback.build(plan, lhs, rhs)
+
+      case _: CartesianProduct =>
+        val lhsPlan = plan.lhs.get
+        val lhsLongCount = pipelines(lhsPlan).numberOfLongs
+        val lhsRefCount = pipelines(lhsPlan).numberOfReferences
+        CartesianProductRegisterPipe(lhs, rhs, lhsLongCount, lhsRefCount, pipeline)(id)
```
Comment (craigtaverner) [inline enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/EnterprisePipeBuilder.scala:248]: The lhsLongCount and lhsRefCount would be known by the lhsCtx inside internalCreateResults, so why do we need to pass it in here?

File: enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/pipes/CartesianProductRegisterPipe.scala [hunk 5]
```diff
@@ -19,18 +19,26 @@
...
+                                       (val id: Id = new Id) extends Pipe {
+
+  protected def internalCreateResults(state: QueryState): Iterator[ExecutionContext] = {
+    lhs.createResults(state) flatMap {
+      lhsCtx =>
+        rhs.createResults(state) map {
+          rhsCtx =>
+            val context = PrimitiveExecutionContext(pipelineInformation)
+            lhsCtx.copyTo(context)
+            rhsCtx.copyTo(context, lhsLongCount, lhsRefCount)
```
Comment (craigtaverner) [inline enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/pipes/CartesianProductRegisterPipe.scala:39]: The lhsLongCount and lhsRefCount would be known by the lhsCtx already, so why do we need to pass it in to the CartesianProductRegisterPipe? Would it not be better to use the CartesianProductPipe pattern of calling `lhsCtx mergeWith rhsCtx` and have the PrimitiveExecutionContext implementation of `mergeWith` understand how to merge two execution contexts like this? Then we could also remove the CartesianProductRegisterPipe class entirely.

**Review by craigtaverner — CHANGES_REQUESTED** [review]

File: enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/pipes/CartesianProductRegisterPipe.scala (hunk 5 above)
Comment (craigtaverner) [reply enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/pipes/CartesianProductRegisterPipe.scala:39]: One issue I see with my idea is that the Cartesian Product pipeline is provided at register allocation time. But it is strictly a concatenation of left and right hand side pipelines, and we are doing that with mergeWith also (at the execution context level). I still feel there are ways to simplify this.

File: enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/EnterprisePipeBuilder.scala (hunk 4 above)
Comment (alexaverbuch) [reply enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/EnterprisePipeBuilder.scala:248]: I agree it's ugly, but even though in the pipe we get a `PrimitiveExecutionContext` it is passed in as the base `ExecutionContext`, so we'd need to do something like 
```
rhsCtx.copyTo(context, lhsCtx.asInstanceOf[PrimitiveExecutionContext].pipeline.numberOfLongs, lhsCtx.asInstanceOf[PrimitiveExecutionContext].pipeline.numberOfReferences)
```

**Review by craigtaverner — APPROVED** [review]

**Pull request merged**